### ISSUE TITLE
use dotenv files for PHP environment variables

### ISSUE THREAD

**Issue by mtmail** [osm-search/Nominatim#946]
Currently we use a PHP file `build/settings/settings.php` (for defaults) plus `local.php` (which the user can change), e.g.

```
<?php
@define('CONST_BasePath', '/home/vagrant/Nominatim');
// General settings
@define('CONST_Debug', false);
@define('CONST_Database_Web_User', 'www-data');
@define('CONST_Max_Word_Frequency', '50000');
```

During installation we create `settings.php` a template and add file paths (in the above example `/home/vagrant/Nominatim`.

https://github.com/vlucas/phpdotenv is the modern alternative.

A year ago melvyn-sopacua started working on that in https://github.com/openstreetmap/Nominatim/issues/634#issuecomment-281333425. Back then we were in the middle of changing the installation (the part that sets the file paths) in preparation for Nominatim 3.0 and we didn't want to introduce more changes.

**Comment by chrisarmitage** [comment]
Hi. I'm an OSM mapping contributor and PHP Developer, and would like to help out with PHP milestone. Whats the best way to find out what need doing and the approach you want to take?

**Comment by mtmail** [comment]
Thanks a lot for the offer! I'll write a more detailed plan.

If you haven't yet it helps to setup a local test environment. The `VAGRANT.md` describes how to set one up in 30 minutes. The `test/README.md` also has a lot of detail how to run tests. https://travis-ci.org/openstreetmap/Nominatim/ has the output of the most recent test runs (see `.travis.yml` what they run). We don't run the `bdd/api` tests on Travis-CI. They require a larger data import, about 11GB. It's a good second or third step but not immediately required. You will notice that more than half of the test suite is actually python (environment.py, queries.py) which in turn execute php-cgi, query.php or the import.

**Comment by mtmail** [comment]
Do you have a base local Nominatim installation?

We still support PHP 5.4 because that's the version the Debian stable release installs. PHPUnit for example no longer supports 5.x releases https://phpunit.de/supported-versions.html 

Our configuration management is `define()` calls in `settings/*.php` files. Luckily we're not affected by PHP 7.3's deprecation of case-insensitive constants https://www.php.net/manual/en/migration73.deprecated.php

We'd like to move `settings/defaults.php` to a `.env.example` file. Users will need to copy it to `.env` during the installation. Optionally they make any local changes to the `.env` file, the usual is setting the database connection string or the base URL.

The key names, e.g. `CONST_HTTP_Proxy_Password` should loose the the `CONST_` prefix. Since those will be over-writable by the shell environment it's standard to uppercase them. So we have `HTTP_PROXY_PASSWORD`. It

One of the advantage of `.env` files is making sure sensitive data (e.g. passwords, auth keys, http proxy) isn't checked into git. With Nominatim we don't have any of those in our repository.

Ideally accessing the confirmation is wrapped in a `Nominatim::Config` so we can enforce certain variables to be set, be of a certain type or set or add helpful warnings. Similar to `Nominatim::DB`.

A lot of modules in `lib/` will need access to a `Nominatim::Config` instance. In the past (db.php until a month ago) we used a global variable. A singleton pattern makes sense but I see that as later optimization.

In `settings/default.php` you will notice `@CMAKE_SOURCE_DIR@` variables. Those are set during the `cmake; make` step of the Nominatim installation. Should be easy to adapt `CMakeList.txt` to do the same to the new files.

The Python test suite writes a `build/settings/local.php` file before testing import of files. See `context.nominatim.local_settings_file` in `test/bdd/steps/queries.py` and `write_nominatim_config` in `test/bdd/steps/environment.py`.

Lastly documentation,`vagrant/*.sh` files and `docs/admin/Migration.md` need to updated.

First step is creating a minimal working prototype.

Confirmation naturally touches a lot of files. In case of Nominatim we also deal with a cmake installation and python test suite. We delayed the project because there was no pressure (no open bugs, no user complaining, no secrets leaked). It's more cleanup, code quality and following https://12factor.net/ best practices.

I can understand if you feel overwhelmed by the many parts. We have listed other open projects, usually larger but less affecting the core Nominatim code, in https://wiki.openstreetmap.org/wiki/Google_Summer_of_Code/2019/Project_Ideas#Nominatim

You can contact me via https://www.openstreetmap.org/user/mtmail/ to exchange email addresses and/or if you have questions.

**Comment by mtmail** [comment]
@chrisarmitage Did you have a look at my write-up and maybe already started some testing locally? I just want to get a sense if you're still planning to work on this milestone. Nominatim v3.3 was released recently and I'm trying to see if this feature could be in 3.4

**Comment by krahulreddy** [comment]
As suggested by @lonvia, I wrote code for this issue, and tested it in `utils/specialphrases.php`. The code added in the file is:
```
require_once(CONST_BasePath.'/lib/Config.php');
$t = new Nominatim\Config();
$t->setEnv();
echo "Using Config Class: ".$t->getEnv("CONST_Phrase_Config")."\n";
```

The .env file is in the Nominatim root directory. The contents of the file is the following during this test:

```
CONST_BasePath=/srv/nominatim/Nominatim

CONST_Phrase_Config="${CONST_BasePath}/settings/phrase_settings.php"
```
This is the Config class that I used.
```
<?php

namespace Nominatim;

use Dotenv\Dotenv as Dotenv;

/**
 * Uses Dotenv to acess environment values
 */
class Config
{
    protected $dotenv;
    
    public function setEnv($dir = __DIR__, $path = "../.env")
    {
        define('DIR_VENDOR', __DIR__.'/../vendor/');
        if (file_exists(DIR_VENDOR . 'autoload.php')) {
            require_once(DIR_VENDOR . 'autoload.php');
            $dotenv = Dotenv::createImmutable($dir, $path);
            $dotenv->load();                
            }
        else
        {
            echo "Failed to load!!";
        }
    }

    public function getEnv($env)
    {
        return getenv($env);
    }


}

```

The output is:
```
Using Config Class: /srv/nominatim/Nominatim/settings/phrase_settings.php
```

Is this the correct way to tackle this issue?

**Comment by lonvia** [comment]
In general that would the right approach. There is one obvious flaw here: `getenv()` looks into the global environment, not something local, so the `$t->getEnv('foo')` could just be replaced by `getenv('foo')` and do the same. I would also prefer the setup you intended and encapsulate the environment in the Config object but as far as I understand dotenv that is not really possible (and also not really wanted, they want it in the global env). Given that situation instantiating the Config class doesn't really make sense. We could just have a static function Nominatim\Config::setup_env() and then continue with `getenv()`.

**Comment by krahulreddy** [comment]
> There is one obvious flaw here: `getenv()` looks into the global environment, not something local, so the `$t->getEnv('foo')` could just be replaced by `getenv('foo')` and do the same. I would also prefer the setup you intended and encapsulate the environment in the Config object but as far as I understand dotenv that is not really possible (and also not really wanted, they want it in the global env).

I agree with this.

> Given that situation instantiating the Config class doesn't really make sense. We could just have a static function Nominatim\Config::setup_env() and then continue with `getenv()`.

I have made the necessary changes.

**Comment by gopi-ar** [comment]
Will these proposed changes support reading from environment variables set in apache / nginx or only from the local `.env` file? These files won't be cached in the opcache and will be read from disk on every request (unless there's a way to?). 

IMO: Deployments will become more complex considering PHP needs env variables to be set at web-server level. The current `settings.php` is optimal for performance since it's cached on first read. If there are concerns about separating config from the repo, the repo can only include a settings.example.php file that can be copied over to a `settings.php` that's part of `.gitignore`.

A relevant note from the dev behind phpdotenv:
> In production, ideally all these variables will already be loaded in the environment though things like Apache's SetEnv directive, etc. PHP's execution model (shared nothing) does not work well with having to read and parse the contents of a file for every single request, thus the reason I say this library was not designed for production, but rather for development, to help you emulate the production environment variables being set.
https://github.com/vlucas/phpdotenv/issues/207#issuecomment-273530307

**Comment by lonvia** [comment]
Fair point, that is a serious issue.

I don't care about the separation, that's reasonably well solved right now. dotenv ticks a number of other boxes:

* language-independent config file (in particular, readable from PHP and Python)
* can pull in configuration from multiple files (aka default vs. local config)
* allows one-time changes to config variables (e.g. changing database just for a single import run). I like uwsgi's approach of allowing every single option to be set as command-line option, in a config file or in the environment. 

Problem is that all approaches that tick those boxes would also have the same issue as the one mentioned in the dotenv tracker. Aka, they would have to parse a file on every call. I'm not sure where that leaves us.

**Comment by gopi-ar** [comment]
When I last looked at Laravel, they work around this by:
- caching the .env file by transforming it into a php file. We'd essentially be generating the `settings.php` from the .env file. We could merge keys from multiple .env files in a specific order (default, then local) & compile it into one `settings.php` file as well. Since the main settings don't change often, we could generate this file before / after installation.
- even with the cache, they'd check to see if the environment had this variable specified & give that preference over the php file - this can solve issue of one-off DB settings by declaring environment for php CLI scripts before executing them (```export ... || php <command>```)

**Comment by lonvia** [comment]
This only affects the website anyway. For the utility scripts the additional work to parse the .env file is not relevant compared to the total execution time. I'd still go for the dotenv approach here.

The website only uses a small subset of all configuration variable. So, yes, we should probably compile a small settings.php for the website scripts. We could even split that and compile a small settings.js as well which receives the settings that affect the HTML output (not that it will survive for a long time in the future but for now it is still there).

**Comment by lonvia** [comment]
Closed by #2115.
